cli/cmd/cluster_prepare.go

```diff
@@ -578,54 +501,257 @@ func runPreflights(ctx context.Context, r *runners, log *logger.Logger, kubeConf
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create registry client")
+	}
+
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(settings.RESTClientGetter(), settings.Namespace(), "secret", golog.Printf); err != nil {
+		return nil, errors.Wrap(err, "init action config")
+	}
+	actionConfig.RegistryClient = registryClient
+
+	client := action.NewInstall(actionConfig)
+	client.ReleaseName = fmt.Sprintf("%s-%d", appSlug, releaseSequence)
+	client.Namespace = r.args.prepareClusterNamespace
+	client.Timeout = r.args.prepareClusterAppReadyTimeout
+	client.Wait = true
+
+	if dryRun {
+		client.DryRun = true
+		client.ClientOnly = true
+	}
+
+	cp, err := client.ChartPathOptions.LocateChart(fmt.Sprintf("oci://%s/%s/release__%d/%s", registryHostname, appSlug, releaseSequence, chartName), settings)
+	if err != nil {
+		return nil, errors.Wrap(err, "locate chart")
+	}
+
+	chartReq, err := loader.Load(cp)
+	if err != nil {
+		return nil, errors.Wrap(err, "load chart")
+	}
+
+	helmRelease, err := client.Run(chartReq, values)
+	if err != nil {
+		return nil, errors.Wrap(err, "run helm install")
+	}
+
+	return helmRelease, nil
+}
+
+func installKotsApp(r *runners, log *logger.Logger, kubeConfig []byte, customer *types.Customer, release *types.AppRelease) error {
+	var releaseYamls []releaseTypes.KotsSingleSpec
+	if err := json.Unmarshal([]byte(release.Config), &releaseYamls); err != nil {
+		return errors.Wrap(err, "failed to unmarshal release yamls")
+	}
+
+	kotsApp, err := kotsutil.GetKotsApplicationSpec(releaseYamls)
+	if err != nil {
+		return errors.Wrap(err, "failed to get kots application")
+	}
+
+	kotsCliVersion := ""
+	if kotsApp != nil && kotsApp.Spec.TargetKotsVersion != "" {
+		kotsCliVersion = kotsApp.Spec.TargetKotsVersion
+	}
+
+	kotsDir, err := os.MkdirTemp("", "kots")
+	if err != nil {
+		return errors.Wrap(err, "failed to create temp dir")
+	}
+	defer os.RemoveAll(kotsDir)
+	kotCLI, err := installKotsCLI(r, kotsCliVersion, kotsDir)
+	if err != nil {
+		return errors.Wrap(err, "failed to install kots cli")
+	}
+
+	license, err := r.api.DownloadLicense(r.appType, r.appID, customer.ID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to download license for customer %q", customer.Name)
+	}
+
+	licenseFile, err := os.CreateTemp(kotsDir, "license")
+	if err != nil {
+		return errors.Wrap(err, "failed to create license file")
+	}
+	defer func() {
+		licenseFile.Close()
+		os.Remove(licenseFile.Name())
+	}()
+	if _, err := licenseFile.Write(license); err != nil {
+		return errors.Wrap(err, "failed to write license file")
+	}
+
+	kubeconfigFile, err := os.CreateTemp(kotsDir, "kubeconfig")
+	if err != nil {
+		return errors.Wrap(err, "failed to create kubeconfig file")
+	}
+	defer func() {
+		kubeconfigFile.Close()
+		os.Remove(kubeconfigFile.Name())
+	}()
+	if err := os.WriteFile(kubeconfigFile.Name(), kubeConfig, 0644); err != nil {
+		return errors.Wrap(err, "failed to write kubeconfig file")
+	}
```

Comment:
Can this be outside `installKotsApp` and `installHelmApp`? Both need a kubeconfig.

Reply:
In the first iteration, I had it in the way you have described.

`installBuilderApp` needs `restKubeConfig` which needs to be built from `kubeConfig []byte`
and it also needs `kubeconfigFlag`.
This would make me pass 2 kubeconfig related parms to function which would not look good.

So, decided to duplicate a bit of code in favour of aesthetics 😉